### PR TITLE
Compatibility with NumPy 1.20+

### PR DESCRIPTION
**Pull request by bnavigator**
* Fix the one already visible failure with NumPy 1.20 (https://github.com/tensorflow/tensorflow/issues/47691#issuecomment-814255904)
* Different approach than https://github.com/tensorflow/tensorflow/pull/48918: **Relax** upper bound instead of pinning to 1.20. Python 3.6 builds still need <1.20. (Per NEP 29, 1.20 dropped Python 3.6)

### CODE REVIEW

**Comment by bhack** [comment]
Just to notify, another user has submitted https://github.com/tensorflow/tensorflow/pull/49008

**Review by mihaimaruseac — APPROVED** [review]

**Comment by gbaned** [comment]
@bnavigator  Can you please resolve conflicts? Thanks!

**Comment by gbaned** [comment]
@bnavigator Can you please resolve conflicts? Thanks!

**Comment by mihaimaruseac** [comment]
The mentioned commit is just a step in getting this handled.

We will keep pinning dependencies in our CI, but `setup.py` would have broader ranges

File: tensorflow/python/ops/array_ops.py
Comment (shoyer) [inline tensorflow/python/ops/array_ops.py:2900]: Do you know what specifically changed in NumPy that broke this?

This may well be a bug in NumPy, but it's hard to say without an isolated reproducer.

Comment (bnavigator) [reply tensorflow/python/ops/array_ops.py:2900]: https://github.com/tensorflow/tensorflow/issues/47691#issuecomment-814255904

**Comment by gbaned** [comment]
@bnavigator Can you please resolve conflicts? Thanks!

**Comment by bnavigator** [comment]
Merged and updated.

So how do I read the logs of the "public" CI, specifically a build with Python >= 3.7, to see whether the correct numpy is used for the build and test?

**Comment by shoyer** [comment]
Unmarking myself as a reviewer here. I'm a Googler but I don't work on TensorFlow.

**Comment by bnavigator** [comment]
@gbaned, please force run on internal again. I think the array_ops change also revealed a minor bug in an unrelated function test. (Possibly also through an updated numpy 1.19 in the internal py3.6 CI (?))

**Review by bhack — APPROVED** [review]
Just to run kokoro

**Comment by bnavigator** [comment]
From the MacOS CPU Python3 fail (Py3.7):

```
  File "/Volumes/BuildData/tmpfs/tmp/bazel/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/kernel_tests/map_ops_test.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 2672, in _GetNdArray
    return np.array(a)
  File "/Volumes/BuildData/tmpfs/tmp/bazel/execroot/org_tensorflow/bazel-out/darwin-opt/bin/tensorflow/python/kernel_tests/map_ops_test.runfiles/org_tensorflow/tensorflow/python/framework/ops.py", line 870, in __array__
    " a NumPy call, which is not supported".format(self.name))
NotImplementedError: Cannot convert a symbolic Tensor (gradient_tape/add/Reshape:0) to a numpy array. This error may indicate that you're trying to pass a Tensor to a NumPy call, which is not supported
```

https://github.com/tensorflow/tensorflow/blob/9c70f6b7aeabeb4eb0dc19a0275bdd17e67412f1/tensorflow/python/framework/test_util.py#L2664-L2673

Something goes wrong in the conversion here. The errors on Windows amount to the same problem.

File: tensorflow/python/ops/array_ops.py
Comment (joshua-cogliati-inl) [reply tensorflow/python/ops/array_ops.py:2900]: What exactly is the parameter ```shape``` here? (As in what are an example type and value for ```shape``` that fails with 1.20 but passes with 1.19?)

Comment (bnavigator) [reply tensorflow/python/ops/array_ops.py:2900]: `symbolic Tensor (gru/strided_slice:0)`

Comment (joshua-cogliati-inl) [reply tensorflow/python/ops/array_ops.py:2900]: I put 5 lines of code that will fail with tensorflow 2.4.1 and numpy 1.21.0 at https://github.com/tensorflow/tensorflow/issues/47691#issuecomment-889506604

Comment (rgommers) [reply tensorflow/python/ops/array_ops.py:2900]: This one line fix is clearly correct, but it looks like this PR cannot be merged because it's changing versions tested in CI (which I agree with maintainers is not necessarily helpful). Why not submit `np.prod` -> `reduce_prod` as a simple one-line fix? The wheels for TensorFlow 2.5.0 on PyPI do not have `<1.19.2`, they only have `~=1.19.2`. So this one-liner fix will unblock users.

Comment (rgommers) [reply tensorflow/python/ops/array_ops.py:2900]: Also looked into this whole mess, and it looks like there is indeed no bug in NumPy. The only thing that changed is https://github.com/numpy/numpy/pull/19001 made a TensorFlow bug (this incorrect `np.prod` usage) more visible.

Comment (bnavigator) [reply tensorflow/python/ops/array_ops.py:2900]: > Why not submit np.prod -> reduce_prod as a simple one-line fix?

TF maintainers would probably do so if they had any interest in supporting newer versions. But IMHO it is very bad practice to supply a fix but not be able to test it. Nobody can guarantee that this one line fix is the only bug that created incompatibility with NumPy >= 1.20

> The wheels for TensorFlow 2.5.0 on PyPI do not have `<1.19.2`, they only have `~=1.19.2`. So this one-liner fix will unblock users.

Nope. `~=1.19.2` [still blocks](https://www.python.org/dev/peps/pep-0440/#compatible-release) `1.20.*` and `1.21.*`

Comment (rgommers) [reply tensorflow/python/ops/array_ops.py:2900]: > Nope. `~=1.19.2` [still blocks](https://www.python.org/dev/peps/pep-0440/#compatible-release) `1.20.*` and `1.21.*`

Argh, tricked again by that funky syntax. Sorry about that.

> Nobody can guarantee that this one line fix is the only bug that created incompatibility with NumPy >= 1.20

There's no guarantee, but the other changes in this PR are not all correct or mergeable. CI should indeed use pinned versions, as was already pointed out. You or someone else could submit a separate PR with only a single CI job change that adds a numpy 1.20.3 test. Now you're changing existing CI jobs that were using 1.18.5 before, and will now jump to 1.20.3. Submitting the fix standalone without changing CI shows that no existing tests are breaking.

> TF maintainers would probably do so

I do agree some more attention seems warranted here. I only got here because of a SciPy bug report where someone could not upgrade because of TensorFlow.

Comment (bnavigator) [reply tensorflow/python/ops/array_ops.py:2900]: > You or someone else could submit a separate PR with only a single CI job change that adds a numpy 1.20.3 test.

Have you reviewed the CI setup of TF? With "internal" and "external", "kokoro", "community build" and so on it is totally unclear to me if it is even possible to add an extra job by a regular PR.

Comment (bnavigator) [reply tensorflow/python/ops/array_ops.py:2900]: So i reverted the changes in the ci_build folder. c996d99

Worst case scenario: This PR now "succeeds" with old NumPy versions, is merged, the issue is closed and nobody ever cares again about actually maintaining TF to support NumPy >= 1.20

Comment (bnavigator) [reply tensorflow/python/ops/array_ops.py:2900]: Note that the comments in #50527 speak about exactly this scenario: We are introducing "fixes" but do not test them.

Comment (rgommers) [reply tensorflow/python/ops/array_ops.py:2900]: It should not close the issue. It just gets the fix that people have been applying by hand merged.

> Have you reviewed the CI setup of TF? With "internal" and "external", "kokoro", "community build" and so on it is totally unclear to me if it is even possible to add an extra job by a regular PR.

yeah it's a bit of a jungle:)

File: tensorflow/tools/pip_package/setup.py
Comment (rgommers) [inline tensorflow/tools/pip_package/setup.py:82]: I'd make this `>=1.19.2,<1.21.2` because it looks like the maintainers do not want to set the version constraint to include not-yet-released versions here.

Comment (rgommers) [reply tensorflow/tools/pip_package/setup.py:82]: Or actually, `<1.22.0` would be better.

File: tensorflow/tools/pip_package/setup.py
Comment (rgommers) [inline tensorflow/tools/pip_package/setup.py:78]: For maintainers: this comment lacks context. Nothing in Python uses semver, including Python itself and TensorFlow. The guidance for NumPy's versioning scheme is at https://numpy.org/devdocs/user/depending_on_numpy.html#understanding-numpy-s-versioning-and-api-abi-stability. The correct upper bounds where you do get backwards compat guarantees is "two minor versions into the future", so right now that is `<1.24.0`.

**Review by rgommers — APPROVED** [review]
The bug fixes here are correct. The upper version limit for NumPy should ideally be `<1.24.0`, but depending on how conservative maintainers want to be it can be set to `<1.22.0` or `<=1.21.1` (the most recent numpy version today).

File: tensorflow/tools/pip_package/setup.py
Comment (bnavigator) [reply tensorflow/tools/pip_package/setup.py:82]: Given the comment above, such a declaration would be only a check that the already installed version is correct. But if it is the only thing that blocks the fix, I am happy do include it.

Comment (rgommers) [reply tensorflow/tools/pip_package/setup.py:82]: Not quite sure what you mean. This ends up in `install_requires`, which will make `pip` respect the constraint:
```python
setup(
   ...
    install_requires=REQUIRED_PACKAGES,
```
The "is already installed" in the comment is confusing/wrong, there is no already-installed version before invoking pip - where could it have come from?

Comment (bnavigator) [reply tensorflow/tools/pip_package/setup.py:82]: If I understand the comment and CI install scripts correctly, this alludes to the fact that you need to install numpy through pip or from other sources (e.g. apt-get) *before* you call  `pip install tensorflow`. Ugly, yes. Welcome to the fun of mixing different requirements resolvers. Nothing to fix within the scope of this PR.

Comment (bnavigator) [reply tensorflow/tools/pip_package/setup.py:82]: Oh, and look at https://github.com/tensorflow/tensorflow/commit/fe2fad5184f2ded836ccfeaa39b557472b69ce50#r52001636 to see where this came from.

Comment (rgommers) [reply tensorflow/tools/pip_package/setup.py:82]: > If I understand the comment and CI install scripts correctly, this alludes to the fact that you need to install numpy through pip or from other sources (e.g. apt-get) before you call pip install tensorflow. Ugly, yes. Welcome to the fun of mixing different requirements resolvers

That's not the case:

```bash
$ conda create -n tf-test
$ conda activate tf-test
$ conda install python=3.8
$ pip install tensorflow
...
numpy                   1.19.5
...
tensorflow              2.5.0
```
Full output of `pip list`:

<details>

```
$ pip list
Package                 Version
----------------------- -------------------
absl-py                 0.13.0
astunparse              1.6.3
cachetools              4.2.2
certifi                 2021.5.30
charset-normalizer      2.0.4
flatbuffers             1.12
gast                    0.4.0
google-auth             1.34.0
google-auth-oauthlib    0.4.5
google-pasta            0.2.0
grpcio                  1.34.1
h5py                    3.1.0
idna                    3.2
keras-nightly           2.5.0.dev2021032900
Keras-Preprocessing     1.1.2
Markdown                3.3.4
numpy                   1.19.5
oauthlib                3.1.1
opt-einsum              3.3.0
pip                     21.2.2
protobuf                3.17.3
pyasn1                  0.4.8
pyasn1-modules          0.2.8
requests                2.26.0
requests-oauthlib       1.3.0
rsa                     4.7.2
setuptools              49.6.0.post20210108
six                     1.15.0
tensorboard             2.5.0
tensorboard-data-server 0.6.1
tensorboard-plugin-wit  1.8.0
tensorflow              2.5.0
tensorflow-estimator    2.5.0
termcolor               1.1.0
typing-extensions       3.7.4.3
urllib3                 1.26.6
Werkzeug                2.0.1
wheel                   0.36.2
wrapt                   1.12.1
```

</details>

There's only one resolver here, namely `pip`'s one.

Comment (rgommers) [reply tensorflow/tools/pip_package/setup.py:82]: > Nothing to fix within the scope of this PR.

Agreed. Unless you wanted to delete the comment, it's just wrong. But I guess only do that if a maintainer asks for it.

Comment (bnavigator) [reply tensorflow/tools/pip_package/setup.py:82]: Ok then as noted in https://github.com/tensorflow/tensorflow/commit/fe2fad5184f2ded836ccfeaa39b557472b69ce50#r52001636 two months ago the comment is completely out of line, but no reaction to my remark, except for  https://github.com/tensorflow/tensorflow/pull/48935#issuecomment-858935936 which does not even address the comment. So I thought I clarify the comment in the setup.py and let the maintainers resolve their inconsitency on their own.

To be honest, I have little motivation left to contribute to TF any further. There is no support from Google or any (open source TF maintainer) to get any progress on this. They can either take the PR or leave it alone.

Comment (rgommers) [reply tensorflow/tools/pip_package/setup.py:82]: Fair enough. Thanks for working on this, much appreciated. Let's see if we get a response here, I would like to see it merged - if only to help the NumPy/SciPy users who want an up-to-date version of both TF and NumPy.

Comment (mihaimaruseac) [reply tensorflow/tools/pip_package/setup.py:82]: > no support form Google..
TF team has been through multiple reorgs over the past year and there has been strategy shifts regarding prioritization of OSS. Happy to say that support is coming soon and we should be able to handle these issues faster.

> up-to-date version of both TF and NumPY
We're working on this, but currently our CI strategy has been to use the exact same set of dependencies in all supported versions of Python and both in CI env and in `setup.py`. We have roadmap to decouple these but it would take a few quarters to fully land, unfortunately.

We can discuss more of this in SIG Build meetings, if needed.

File: tensorflow/tools/pip_package/setup.py
Comment (mihaimaruseac) [reply tensorflow/tools/pip_package/setup.py:78]: You are right, the versioning scheme is similar to semver (compared to PVP for example) but the backwards compat guarantees are less severe.

We have a roadmap at the end of which we would be able to use a wide range of dependencies in `setup.py` after testing that TF works with any dependency version in the range. But this would take a significant amount of work that has just recently been prioritized

File: tensorflow/tools/pip_package/setup.py
Comment (rgommers) [reply tensorflow/tools/pip_package/setup.py:82]: Thanks @mihaimaruseac, great to hear more priority will be given to this. I don't know what SIG Build meetings are, but please feel  free to ping me on NumPy versioning and wheels/conda dependency versioning topics.

File: tensorflow/tools/pip_package/setup.py
Comment (bnavigator) [inline tensorflow/tools/pip_package/setup.py:79]:
```suggestion
```

File: tensorflow/tools/pip_package/setup.py
Comment (bnavigator) [inline tensorflow/tools/pip_package/setup.py:82]:
```suggestion
    'numpy >=1.19.2, <1.22.0',
```

Comment (mihaimaruseac) [reply tensorflow/tools/pip_package/setup.py:82]: bit.ly/tf-sig-build-notes has all the details about SIG Build. Basically, it's a group of people interested in building TF on different environments

**Review by mihaimaruseac — APPROVED** [review]

File: tensorflow/python/ops/array_ops.py
Comment (mihaimaruseac) [reply tensorflow/python/ops/array_ops.py:2900]: I'll make sure that by the end of the year we'll have support for wide range of dependencies.

**Review by mihaimaruseac — APPROVED** [review]

**Comment by casparvl** [comment]
I have build TensorFlow 2.5.0 from sources against `numpy 1.20.3`. Then, I ran the TF test suite and ran into the issue described [here](https://github.com/tensorflow/tensorflow/issues/47691#issuecomment-814255904) (8 tests from the test suite were failing). I've applied the patches suggested in this PR, rebuilt TF 2.5.0 against the same numpy, and reran the TF test suite.

The tests that were failing on the `np.prod(...)` line are indeed fixed by the changes this PR makes to tensorflow/python/ops/array_ops.py. So far so good.

Now, only 2 tests are still failing:

```
ERROR: testLookupAddGrad (__main__.MapOpsTest)
MapOpsTest.testLookupAddGrad
----------------------------------------------------------------------
Traceback (most recent call last):
...
  File "/scratch/casparl/TensorFlow/2.5.0/foss-2021a-CUDA-11.3.1/tmplnPLAE-bazel-tf/828a44b36155f75b299651f28995fd3c/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/kernel_tests/map_ops_test.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 2586, in _GetNdArray
    return np.array(a)
  File "/scratch/casparl/TensorFlow/2.5.0/foss-2021a-CUDA-11.3.1/tmplnPLAE-bazel-tf/828a44b36155f75b299651f28995fd3c/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/kernel_tests/map_ops_test.runfiles/org_tensorflow/tensorflow/python/framework/ops.py", line 867, in __array__
    raise NotImplementedError(
NotImplementedError: Cannot convert a symbolic Tensor (gradient_tape/add/Reshape:0) to a numpy array. This error may indicate that you're trying to pass a Tensor to a NumPy call, which is not supported

======================================================================
ERROR: testLookupMultiplyGrad (__main__.MapOpsTest)
MapOpsTest.testLookupMultiplyGrad
----------------------------------------------------------------------
Traceback (most recent call last):
...
    a_as_ndarray = self._GetNdArray(a)
  File "/scratch/casparl/TensorFlow/2.5.0/foss-2021a-CUDA-11.3.1/tmplnPLAE-bazel-tf/828a44b36155f75b299651f28995fd3c/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/kernel_tests/map_ops_test.runfiles/org_tensorflow/tensorflow/python/framework/test_util.py", line 2586, in _GetNdArray
    return np.array(a)
  File "/scratch/casparl/TensorFlow/2.5.0/foss-2021a-CUDA-11.3.1/tmplnPLAE-bazel-tf/828a44b36155f75b299651f28995fd3c/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/python/kernel_tests/map_ops_test.runfiles/org_tensorflow/tensorflow/python/framework/ops.py", line 867, in __array__
    raise NotImplementedError(
NotImplementedError: Cannot convert a symbolic Tensor (gradient_tape/cond_1/Identity:0) to a numpy array. This error may indicate that you're trying to pass a Tensor to a NumPy call, which is not supported
```
Essentially, this is what @bnavigator also mentioned in [this comment](https://github.com/tensorflow/tensorflow/pull/48935#issuecomment-863264915)

I'm wondering, @bnavigator , did you figure out what the solution is there? I'm not really deep enough into the TensorFlow & numpy codes to understand why this conversion would fail, so I'm not sure how to fix it...

**Comment by Flamefire** [comment]
@bnavigator 
> Something goes wrong in the conversion here. The errors on Windows amount to the same problem.

The issue is due to this change in numpy 1.20: https://numpy.org/doc/stable/release/1.20.0-notes.html#arraylike-objects-which-do-not-define-len-and-getitem

What we have here is a list of TF tensors: `[a, b]` where the Tensor has the `__array__` method which is (wrongly) called now and it also has a `__len__` but no `__getitem__` method. --> Fix would likely be to add the latter

**Comment by bnavigator** [comment]
Both errors occur in a piece of code that tries to convert something into an `np.ndarray`. So you need to make sure that this piece actually results in `a` being an actual `ndarray`:

https://github.com/tensorflow/tensorflow/blob/573c905ff7b96d86d8deb5e3955ef04a8e738370/tensorflow/python/framework/test_util.py#L2679-L2683


That should be the case for both  `tensorflow.python.framework.ops.EagerTensor` and `tensorflow.python.framework.ops.Tensor`

**Comment by Flamefire** [comment]
@bnavigator The problem is that `a` is a list which contains a `Tensor` and numpy 1.20+ recursively converts sequences into numpy arrays. I.e. you have `np.array([Tensor(), [1,2,3]])` due to https://github.com/tensorflow/tensorflow/blob/573c905ff7b96d86d8deb5e3955ef04a8e738370/tensorflow/python/framework/test_util.py#L2684-L2685

I can confirm that removing the "unimplemented" `__array__` function fully solves all issues without changing anything else (even without this whole PR)

**Comment by bnavigator** [comment]
> numpy 1.20+ recursively converts sequences into numpy arrays

Before that happens, you call `np.numpy(something)` and that `something` contains a Tensor. Which is not allowed.
 
> I can confirm that removing the "unimplemented" `__array__` function fully solves all issues without changing anything else (even without this whole PR)

As I understand, removing it discards the safeguard **You shalt not pass a Tensor to a NumPy function**. Not as a single value nor as part of a sequence.

Instead, you have to make sure that either `_GetNdArray(a)` is only called with Tensor-likes that are convertible or add extra logic to break down sequences and return them in a multidimensional `ndarray`.

**Comment by Flamefire** [comment]
> Before that happens, you call np.numpy(something) and that something contains a Tensor. Which is not allowed.

It was in pre numpy 1.20 where it returns `np.array(Tensor, [1,2,3])`, i.e. non-recursive conversion

> As I understand, removing it discards the safeguard You shalt not pass a Tensor to a NumPy function. Not as a single value nor as part of a sequence.

More or less true. At the moment (i.e. in current TF releases) passing it as a sequence is allowed de facto. With numpy 1.20 it is no longer allowed and TF code would need to adapt to that.

>  add extra logic to break down sequences and return them in a multidimensional ndarray.

That is likely true. Or at least use a `TypeError` from `__array__` which is handled by the except clause.

However to me adding a `__array__` function which (always) throws an error is a violation of the protocol. But not sure here...

**Comment by bnavigator** [comment]
Could you please test with af47a19cd50477834bc54f8351c382f3a74e17ee?

File: tensorflow/python/framework/test_util.py
Comment (NeilGirdhar) [inline tensorflow/python/framework/test_util.py:2689]: Is this really better?  What if a user calls `np.array` on such a sequence?  I think the idea of removing the bad `__array__` method is superior.

**Comment by rgommers** [comment]
> However to me adding a `__array__` function which (always) throws an error is a violation of the protocol. But not sure here...

It's not, raising an error from `__array__` is a valid thing to do to prevent numpy from doing anything unwanted, such as treating a Tensor as a sequence, or converting it to a size-1 object array.

File: tensorflow/python/framework/test_util.py
Comment (bnavigator) [reply tensorflow/python/framework/test_util.py:2689]: IMHO, yes it is better. Calling `np.array` on a sequence with a tensor in it is plain wrong ("You shalt not ...") and a user should get a proper error message.

This merely catches the case where we accidentally did it in test_util and try a different approach.

**Comment by Flamefire** [comment]
> It's not, raising an error from __array__ is a valid thing to do to prevent numpy from doing anything unwanted, such as treating a Tensor as a sequence, or converting it to a size-1 object array.

Ok, yes you are right. Then I'd rather add the NotImplementedError to https://github.com/tensorflow/tensorflow/blob/573c905ff7b96d86d8deb5e3955ef04a8e738370/tensorflow/python/framework/test_util.py#L2793 or change it to a Type/ValueError in `__array__`

File: tensorflow/python/eager/function_test.py
Comment (Flamefire) [inline tensorflow/python/eager/function_test.py:289]: This change looks unrelated

**Comment by bnavigator** [comment]
> Ok, yes you are right. Then I'd rather add the NotImplementedError to
> 
> https://github.com/tensorflow/tensorflow/blob/573c905ff7b96d86d8deb5e3955ef04a8e738370/tensorflow/python/framework/test_util.py#L2793
> 
> or change it to a Type/ValueError in `__array__`

I don't understand `__array__` throws a `NotImplementedError` so I have to catch that. I cannot change one OR the other. Only both. Which would be an API change.

**Comment by Flamefire** [comment]
> I don't understand __array__ throws a NotImplementedError so I have to catch that. I cannot change one OR the other. Only both. Which would be an API change.

No. You have to change exactly one.

IMO The error here is not "not implemented" but rather: "not valid on this type/value" and hence should be a TypeError/ValueError in `__array__`. With that change the linked except will catch that and do the correct thing.
If one does not want to change the "API" (throw another error), then NotImplementedError should be added to the except list.

**Comment by rgommers** [comment]
`TypeError` seems correct to me.

**Comment by bnavigator** [comment]
I didn't introduce the `NotImplementedError` in `Tensor.__array__()` and I don't want to override the developer's choice here. Looks like they might want to implement addionial functionality in the future:

https://docs.python.org/3/library/exceptions.html#TypeError

> This exception may be raised by user code to indicate that an attempted operation on an object is not supported, and is not meant to be. If an object is meant to support a given operation but has not yet provided an implementation, NotImplementedError is the proper exception to raise.

https://github.com/tensorflow/tensorflow/blob/305a8ba41978e7954c1a13baebc94bdb8ad16514/tensorflow/python/framework/ops.py#L898-L912

**Comment by Flamefire** [comment]
Your quote supports the claim that TypeError should be the correct error: A symbolic Tensor (i.e. this type) cannot be converted

**Comment by bnavigator** [comment]
> Ok, yes you are right. Then I'd rather add the NotImplementedError to
> 
> https://github.com/tensorflow/tensorflow/blob/573c905ff7b96d86d8deb5e3955ef04a8e738370/tensorflow/python/framework/test_util.py#L2793
> 
> or change it to a Type/ValueError in `__array__`

Okay, after expanding the code snipped around the single referenced line, now I undestand.

**Comment by Flamefire** [comment]
> I don't think it is appropriate to change this in a PR titled "Compatibility with NumPy 1.20+"

As mentioned: it is one way to fix this issue, unless that error really has to be a NotImplementedError, which I doubt. As you agreed the TypeError would be more suitable here, so fixing the (more or less) wrong error raised will also fix the numpy 1.20 issue.
But only the TF guys can tell that for sure.

**Comment by bnavigator** [comment]
@robieta #30694 ebca088d52c135551268717b34d79c47d04e6794

**Comment by Flamefire** [comment]
FWIW: A similar thing is happening for KerasTensor and there `TypeError` is used. I'd hence say changing the error is correct: https://github.com/tensorflow/tensorflow/blob/305a8ba41978e7954c1a13baebc94bdb8ad16514/tensorflow/python/keras/engine/keras_tensor.py#L248

**Comment by bnavigator** [comment]
I suggest, you make a PR and if it is merged before this one, I will happily revert https://github.com/tensorflow/tensorflow/pull/48935/commits/9a4642fb897e8ac7901253a4225d0d346715545e.

**Comment by bnavigator** [comment]
Could someone please trigger a kokoro run? @bhack, @mihaimaruseac?

**Comment by mihaimaruseac** [comment]
Sorry, I was OOO. It seems now we have wide ranges of deps in setup.py and this causes conflicts.

**Comment by bnavigator** [comment]
Merge conflict resolved.

**Review by mihaimaruseac — APPROVED** [review]

**Comment by svenstaro** [comment]
Can't this be closed now according to #52380?

**Comment by bnavigator** [comment]
Yeah, https://github.com/tensorflow/tensorflow/pull/47957 "fixed" the `np.prod` error between TF 2.5 and 2.6 by catching the Exception correctly and not returning a constant. 0f8fde4 changes it to not run into the exception.

The changes in the requirements specifiers, the wrong comment about numpy install order and the wrong float shape tuples are still something to consider.

**Comment by svenstaro** [comment]
Alright, fair enough.

**Comment by wei-v-wang** [comment]
Please consider reproducers in https://github.com/tensorflow/tensorflow/issues/52657 closed this as this PR is supposed to address such issues.

**Comment by mihaimaruseac** [comment]
We now merged most of this PR. Disabled tests will be fixed ahead in the near future.

**Pull request closed without merging**
